### ISSUE TITLE
Should we consider automatically flagging posts which hit more than one filter?

### ISSUE THREAD

**Issue by angussidney** [Charcoal-SE/SmokeDetector#339]
As we all know, the role of the SmokeDetector project is to detect and delete spam as fast as possible. Of course, the fastest way to delete spam is to have as many people flag it as fast as possible.

So, **what if we made Smokey automatically flag posts which hit more than one filter**, so that spam gets deleted faster? [According to Undo](http://chat.stackexchange.com/transcript/message/33903186#33903186), reports which hit more than one filter are pretty damn accurate: 

 - At least two reasons: 22136 TPs, 424 FPs (~2% false positives)
 - At least three reasons: 13087 TPs, 24 FPs (~0.2% false positives)

When you compare that to the helpful flag percentage of a human ([like me](https://i.stack.imgur.com/aCdsq.png)) of ~95%, Smokey is definitely accurate enough for us to consider automated flagging.

Of course, we can do more to make the auto-flags more accurate:

 - Increase the number of filters required before an auto-flag is cast
 - Exclude less accurate reasons from the count of filters hit, such as Link at End of Answer, Repeating characters in body etc.
 - Revert flags if FP feedback is provided (is this possible via the API? Maybe a Meta FR?)

However, there will be some things we need to think about before we do anything:

 - NAA posts need to be separated from red-flaggable posts
 - Do we need to separate spam and offensive posts (well, yes.... but will one incorrect type of red flag by smokey make much of a difference in comparison to 5 correct humans)?
 - Would SE be happy with us doing this? (I assume the answer would be yes, since they are already considering integration with us at our current accuracy, and with a little tweaking we can make auto-flags based on the number of filters more accurate than Andy's auto-comment-flagging, which everyone seems to be happy about)

*But Angus, isn't that a huge amount of effort for just one extra flag? Surely it can't make much of a difference compared to our current flagger userbase?*

Well, yes, in some ways, you are right. Most of the time, it won't make much of a difference, but it certainly will make spam get deleted faster, especially on smaller sites, where every flag counts. 

Also, now that SE is starting to talk to us about the possibility of integration, I think it is important that we have a tried-and-tested way of identifying which posts are almost definitely spam, which we can present to them. Our overall false positive rate of 17% isn't good enough to be put into production on one of the biggest sites on the internet, so I think we should start work on a more accurate identification system so that it is ready for when SE needs it.

---

**So, what does everyone else think? Please share your thoughts, ideas, and criticizms.**

Of course, this is only an initial basic idea. Please don't nitpick at the specifics - if (mostly) everyone agrees to this, we can iron out any details in another discussion. For now, let's just discuss whether we want to put some serious effort into this.

**Comment by AWegnerGitHub** [comment]
If we are automating flags and especially spam flags, which carry severe penalties, we need to be more accurate than 2% false positives. I am more comfortable with the 0.2% false positive rate though that 3 reasons seems to have.

Regarding ways to make it more accurate, I think the following should be considered:

 - Don't flag NAA posts. 
 - Not flagging vandalism posts. These can be fixed by rolling back the edit. A spam flag isn't appropriate. We'd need to be able to accurately detect a post has been vandalized though.
 - Apply a set of weights to our various reasons. If a threshold is met/passed, then flag the post. This allows us to fine tune specific reasons as more or less relevant to a post being spam. It also lets us test new reasons without potentially throwing accuracy off (the new reason could have a 0 weight). 
 
As @Art mentioned in his post, we have discussed this previously. If we auto flag, we can't nuke these from orbit using only Smokey and users that want to autoflag. That has a high potential of accidentally nuking something that shouldn't be within seconds. 

One last point, which is a minor technical detail, Smokey (or any users that are being used to automatically flag) will need to be kept updated with accounts on all sites.

**Comment by Undo1** [comment]
I'm fairly confident we can drop the rate further below 0.2%, which is already a heck of a lot better than any human. I'm at 96.8% on spam flags on SO.

**Comment by honnza** [comment]
Vandalism detector is easy: just check if there's been an edit on the post. Unless the API doesn't reveal that information?
+1 on reason weights. How do we pick the weights, though?

**Comment by AWegnerGitHub** [comment]
@honnza - Experience with which reasons are more accurate and trial and error. The weights would have to be adjusted over time to find good values.

**Comment by Undo1** [comment]
We don't get revision information from the API requests Smokey makes, but we would definitely have enough quota on the metasmoke side to get all the information we want if we're considering flagging something

**Comment by honnza** [comment]
Worst case you could spawn more bots = more quota

**Comment by ArtOfCode-** [comment]
No, that'd be abuse of the API. One key, one instance per app.

**Comment by SmokeDetector** [comment]
Or just cycle API keys, but we shouldn't need to do that anyway.

**Comment by honnza** [comment]
I meant, grab a second key.

**Comment by ArtOfCode-** [comment]
I know, but using more than one key per app just to get more quota is abuse of the API. I'd want to talk to SE before we did anything like that.

**Comment by honnza** [comment]
That wouldn't be the same app either. Instead, multiple apps with multiple
keys.

**Comment by Undo1** [comment]
Sure @Art, it'd probably be a "hey can we get more quota" "sure but we
forgot where the button is to make that happen, just use two keys" kind of
thing. I highly doubt there'd be objections to this project using 20k a
day.

**Comment by dorukayhan** [comment]
Who says we can't focus on *part of a filter*?

For example, the bot could check if an answer posted by a user called "wingding" matches `bojiter|kimochi` and flag the answer as abusive if it does match.

**Comment by dorukayhan** [comment]
Alright, here are two combinations we can start with:

- [Pattern-matching website in a question on Drupal Answers](https://metasmoke.erwaysoftware.com/search?utf8=%E2%9C%93&title=&body=&username=&why=&site=86&feedback=&reason=35&user_rep_direction=%3E%3D&user_reputation=0&commit=Search) (no false positives)
- [Blacklisted website and a link at end in the same question](https://metasmoke.erwaysoftware.com/search?utf8=%E2%9C%93&title=&body=&username=&why=Post+-+Link+at+end%3A&site=&feedback=&reason=22&user_rep_direction=%3E%3D&user_reputation=0&commit=Search) ([only one false positive 7 months ago](https://metasmoke.erwaysoftware.com/post/27511)

As you see, not only these combinations catch a lot of spam nowadays but also they have an insane TP rate. In fact, I'm confident in "pattern-matching website on Drupal.SE" enough that I'd support creating a sockpuppet just to raise *two* autoflags on each Drupal spam.

**Comment by Undo1** [comment]
Yeah, I need to bug some Stack folks (or maybe just one) to give them a definitive chance to veto this. Otherwise, I don't see why we can't start dry running this - build the infrastructure, just don't actually flag stuff.

We need to log everything around this and be accountable for everything. Giving mods a heads up before enabling this on their site would probably be okay too, otherwise we're going to be fielding a ton of questions about this.

In short, it's a pretty big operation and we probably need to just dive into it.

**Comment by honnza** [comment]
I'm not sure a flagging sock would be appreciated by the dev team, but I do
approve of the choice of the heuristics

**Comment by SmokeDetector** [comment]
If we were to do this, I'd be happy to run it under Smokey's account or my
account. I don't want to do anything automatically that I wouldn't do with
my own account.

**Comment by angussidney** [comment]
Everyone seems to be discussing various filters which we could include, but I'd like remind you of my original proposal: auto-flag posts which hit 3 or more filters. At only 0.2% false positives, it's extremely accurate, but also identifies a lot of posts (about 1 in 4 posts), compared to the relatively small pockets that we can identify by selecting reasons manually.

**Comment by markhurd** [comment]
I wonder if you could act like humans: one bot per rule, and only post the flag if a human has already flagged as spam, and the bots communicate amongst themselves and also post their flag if at least two of them are ready to do so.

**Comment by Undo1** [comment]
@markhurd This is centralized anyway, there aren't really multiple copies of Smokey running around. It's probably a decent idea to wait for positive feedback before flagging, at least at first.

**Comment by angussidney** [comment]
It has begun!

https://chat.stackexchange.com/rooms/11540/conversation/babys-first-flag
